tests/test/subtype.rs

```diff
@@ -121,6 +152,160 @@ fn multi_lifetime() {
+            // }]"
+            //
+            // This is incorrect, as we shouldn't be requiring 'a and 'b to be
+            // related to eachother. Instead, U should be &'?1 u32, with constraints
+            // ?1 : 'a, ?1: 'b.
+            "Unique; for<?U1> { substitution [?0 := (&'^0.0 Uint(U32))], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '!1_0: '^0.0  }, \
+            InEnvironment { environment: Env([]), goal: '!1_1: '^0.0  }] }"
+        }
+    }
+}
+
+/// Tests that the generalizer generalizes lifetimes within a covariant struct.
+#[test]
+fn multi_lifetime_covariant_struct() {
+    test! {
+        program {
+            #[variance(Covariant)]
+            struct Foo<A> {}
+        }
+
+        goal {
+            forall<'a, 'b> {
+                exists<U> {
+                    Subtype(Foo<&'a u32>, Foo<U>),
+                    Subtype(Foo<&'b u32>, Foo<U>)
+                }
+            }
+        } yields {
+            // Result should be identical to multi_lifetime result.
+            "Unique; for<?U1> { substitution [?0 := (&'^0.0 Uint(U32))], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '!1_0: '^0.0  }, \
+            InEnvironment { environment: Env([]), goal: '!1_1: '^0.0  }] }"
+        }
+    }
+}
+
+/// Tests that the generalizer generalizes lifetimes within a contravariant
+/// struct, and that that variance contravariance is recognized.
+#[test]
+fn multi_lifetime_contravariant_struct() {
+    test! {
+        program {
+            #[variance(Contravariant)]
+            struct Foo<A> {}
+        }
+
+        goal {
+            forall<'a, 'b> {
+                exists<U> {
+                    Subtype(Foo<&'a u32>, Foo<U>),
+                    Subtype(Foo<&'b u32>, Foo<U>)
+                }
+            }
+        } yields {
+            // Result should be opposite multi_lifetime result.
+            "Unique; for<?U1> { substitution [?0 := (&'^0.0 Uint(U32))], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '^0.0: '!1_0 }, \
+            InEnvironment { environment: Env([]), goal: '^0.0: '!1_1 }] }"
+        }
+    }
+}
+
+/// Tests that the generalizer generalizes lifetimes within a invariant struct.
+#[test]
+fn multi_lifetime_invariant_struct() {
+    test! {
+        program {
+            #[variance(Invariant)]
+            struct Foo<A> {}
+        }
+
+        goal {
+            forall<'a, 'b> {
+                exists<U> {
+                    Subtype(Foo<&'a u32>, Foo<U>),
+                    Subtype(Foo<&'b u32>, Foo<U>)
+                }
+            }
+        } yields {
+            // Because A is invariant, we require the lifetimes to be equal
+            "Unique; substitution [?0 := (&'!1_1 Uint(U32))], lifetime
+            constraints [InEnvironment { environment: Env([]), goal: '!1_0: '!1_1 }, \
+            InEnvironment { environment: Env([]), goal: '!1_1: '!1_0 }]"
```

Comment:
Hmm, I wonder if it would be better to introduce a new variable `?U1` where both `'!1_0` and `!1_1` are equal. I guess, logically, they are the same; and this is what we get from short-circuiting the `Invariant` case.